codetransformer/patterns.py

```diff
@@ -0,0 +1,289 @@
+from operator import methodcaller, index
+import re
+from types import MethodType
+
+from .utils.instance import instance
+from .utils.immutable import immutable
+
+
+mcompile = methodcaller('mcompile')
+
+
+def _prepr(m):
+    if isinstance(m, or_):
+        return '(%r)' % m
+
+    return repr(m)
+
+
+def coerce_ellipsis(p):
+    """Convert ... into a matchany
+    """
+    if p is ...:
+        return matchany
+
+    return p
+
+
+class matchable:
+    """Mixin for defining the operators on patterns.
+    """
+    def __or__(self, other):
+        other = coerce_ellipsis(other)
+        if self is other:
+            return self
+
+        if not isinstance(other, matchable):
+            return NotImplemented
+
+        patterns = []
+        if isinstance(self, or_):
+            patterns.extend(self.matchables)
+        else:
+            patterns.append(self)
+        if isinstance(other, or_):
+            patterns.extend(other.matchables)
+        else:
+            patterns.append(other)
+
+        return or_(*patterns)
+
+    def __ror__(self, other):
+        # Flip the order on the or method
+        return type(self).__or__(coerce_ellipsis(other), self)
+
+    def __invert__(self):
+        return not_(self)
+
+    def __getitem__(self, key):
+        try:
+            n = index(key)
+        except TypeError:
+            pass
+        else:
+            return matchnm(self, n)
+
+        if isinstance(key, tuple) and len(key) in (1, 2):
+            return matchnm(self, *key)
+
+        if not isinstance(key, modifier):
+            raise TypeError('invalid modifier: {0}'.format(key))
+
+        return modified(self, key)
+
+
+class modified(immutable, matchable):
```

Comment:
`postfix_modifier` maybe?  There are lots of logical modifiers here that aren't subclasses of this type.